subprojects/core/src/main/java/org/gradle/api/internal/cache/DefaultCacheConfigurations.java

```diff
@@ -127,14 +136,38 @@ public void synchronize(CacheConfigurationsInternal persistentCacheConfiguration
         persistentCacheConfigurations.getDownloadedResources().getRemoveUnusedEntriesOlderThan().value(getDownloadedResources().getRemoveUnusedEntriesOlderThan());
         persistentCacheConfigurations.getCreatedResources().getRemoveUnusedEntriesOlderThan().value(getCreatedResources().getRemoveUnusedEntriesOlderThan());
         persistentCacheConfigurations.getCleanup().value(getCleanup());
+        persistentCacheConfigurations.getMark().value(getMark());
+    }
+
+    public void finalizeConfiguration(Gradle gradle) {
+        finalizeConfigurationValues();
+        if (getMark().get()) {
+            markCacheDirectories(gradle);
+        }
     }
 
-    public void finalizeConfigurationValues() {
+    private void finalizeConfigurationValues() {
         releasedWrappersConfiguration.getRemoveUnusedEntriesOlderThan().finalizeValue();
         snapshotWrappersConfiguration.getRemoveUnusedEntriesOlderThan().finalizeValue();
         downloadedResourcesConfiguration.getRemoveUnusedEntriesOlderThan().finalizeValue();
         createdResourcesConfiguration.getRemoveUnusedEntriesOlderThan().finalizeValue();
         getCleanup().finalizeValue();
+        getMark().finalizeValue();
+    }
+
+    private static void markCacheDirectories(Gradle gradle) {
+        CacheDirUtil.tryMarkCacheDirectory(gradle.getGradleUserHomeDir().toPath().resolve(
+            WrapperDistributionCleanupAction.WRAPPER_DISTRIBUTION_FILE_PATH
+        ));
+        CacheDirUtil.tryMarkCacheDirectory(gradle.getGradleUserHomeDir().toPath().resolve(
```

Comment:
comment: I'm not sure there is any real danger of these changing in the future, but it makes me slightly uncomfortable that these are decoupled from the things that create each of these directories.  Feels like we should have a service (or services) in `GlobalScopeServices` that is a single source of truth and provides these directories to everything in the build that needs to know about them.  The "daemon" dir is probably a little tricky, but it seems like the other two would be easy to capture.

I don't think we need to do anything about this now unless you want to.

Reply:
Yes, I would like this too.